Wait for TaskQueue to be ready before calling wrapper chain

jakobwesthoff opened the pull request:
If we are not waiting for the TaskQueue, of the currently active pouchdb instance to be ready, before calling the handler, we might end up calling it twice. Once due to the call from the outside, which initiated the chain and once from the queuing itself, which calls it again, once the pouchdb instance is ready.

This race condition does not always happen, but I found it to happen more often with PouchDB >6.1.1 and the idb-adapter. This is most likely due to optimizations in the setup process of this adapter.

jakobwesthoff committed:
Wait for TaskQueue to be ready before calling wrapper chain

If we are not waiting for the TaskQueue, of the currently active pouchdb instance to be ready, before calling the handler, we might end up calling it twice. Once due to the call from the outside, which initiated the chain and once from the queuing itself, which calls it again, once the pouchdb instance is ready.

This race condition does not always happen, but I found it to happen more often with PouchDB >6.1.1 and the idb-adapter. This is most likely due to optimizations in the setup process of this adapter.

gr2m commented:
@nolanlawson @daleharvey @garrensmith @marten-de-vries can one of you merge and release the new version? Or add me as npm collaborator to the package?

This looks solid to me, it goes together with https://github.com/pouchdb-community/transform-pouch/pull/46#pullrequestreview-51914870

gr2m commented on index.js:
```diff
@@ -443,8 +443,28 @@ function callHandlers(handlers, args, method) {
+    });
+  } else {
+    promise = method();
+  }
```
another pattern and maybe more readable would be 

```js
var promise = Promise.resolve()
if (!args.base.taskqueue.isReady) {
  promise = promise.then(function () {
    new Promise(function (resolve, reject) {
      args.base.taskqueue.addTask(function (failed) {
        if (failed) {
          return reject(failed);
        }

        resolve();
      });
    });
  })
}

promise = promise.then(function () {
  return method()
})
```

jakobwesthoff replied:
You are right. I usually use this pattern as well. I have got no idea, why I didn't do it here. Should I change it and push a new commit, or do you want to stay with the original implementation?

gr2m replied:
if you agree that it’s more readable, send another commit, yeah. We can squash & merge the commits once we merge. Thanks

jakobwesthoff committed:
Used different pattern to handle conditional async operation

jakobwesthoff commented:
Requested changes added.

jakobwesthoff committed:
Revert "Used different pattern to handle conditional async operation"

This reverts commit 7e1069530d626a7f68890016da14377a75deb114.

jakobwesthoff commented:
I reverted the commit, when I realized it doesn't behave correctly in correspondence with the application I am using it in. The original version seems to work quite fine. I am not sure, where the difference between the two implementations is. (Expect a missing `return` in line 453, which I already fixed while testing)

However I wanted to make sure no "broken" version is merged.

The given one does function properly, so I would suggest to stay with it, as I currently don't have the time to find out why the new version is not identical to the first one.

garbados commented:
Howdy!

This bug is still a huge problem! Specifically it shows up in the test suite for `transform-pouch` via this issue: https://github.com/pouchdb-community/transform-pouch/issues/55

It looks like the Travis CI build failed over Ruby arcana. Is there anything a passerby can do to help this PR get merged?

janl commented:
Turns out this repo isn’t really used in transform-pouch (where we need it), but there is a variant of this published as part of pouchdb-server (https://github.com/pouchdb/pouchdb-server/tree/master/packages/node_modules/pouchdb-wrappers).
